Fix losing physics after setting pose

songxiaocheng opened the pull request:
Fixes #1272
Fixes #1856
Fixes #1246
Fixes #3049 : After using `simSetVehiclePose` in multirotor mode, the drone can't be controlled by most of the vehicle commands.
It is because when starting playing, the physical body is set to be grounded. Then we set pose directly, but the `body.isGrounded()` is still true, which causes `FastPhysicsEngine` not working as expected.

laurelkeys commented:
Hey @songxiaocheng I've tried your solution but there's still some unexpect behavior when calling `simSetVehiclePose` consecutive times, the drone seems to always start flying back to the first position it teleported to.

Could you test this short example on `Blocks` to check if the same problem occurs for you?
```python
ORIENTATION = airsim.Quaternionr(0.012, -0.156, 0.076, 0.985)

POSITIONS = [(15, 18, -40),  (55, 18, -40), (55, -20, -40), (15, -20, -40), (15, 18, -40)]

client = airsim.MultirotorClient()
client.confirmConnection()
client.enableApiControl(True)
client.armDisarm(True)

try:
    if client.getMultirotorState().landed_state == airsim.LandedState.Landed:
        client.takeoffAsync(timeout_sec=5).join()
    client.hoverAsync().join()  

    for position in POSITIONS:
        pose = airsim.Pose(airsim.Vector3r(*position), ORIENTATION)
        client.simSetVehiclePose(pose, ignore_collison=True)
        client.hoverAsync().join()
        time.sleep(2)

    client.landAsync().join()
except KeyboardInterrupt:
    client.reset()
```

Any clues of why this happens/how to fix it would be much appreciated 🙂

p.s. I've also tried things like `client.moveByManualAsync(0, 0, 0, duration=2).join()` instead of `time.sleep(2)` without success, so that doesn't seem to be the cause of the problem.

songxiaocheng commented:
Hi @laurelkeys, thanks for your feedback, and I just figured out where the problem is.

By the way, as the author said, [the `simSetVehiclePose` is meant for CV mode](https://github.com/microsoft/AirSim/issues/1272#issuecomment-407269765). So as not to mislead others, the above short example is only for debugging, rather than the conventional way to use Airsim. However, I think `simSetVehiclePose` is also pretty useful for non-CV mode if it works well.

At first, there are two facts in the current implementation:

- The `hover` command eventually uses `moveOnPath(path,...)` function, where the input `path` contains only one point, that is the current point. However, since the drone is already at the current point, the function `moveOnPath` invokes NO control command. See `MultirotorApiBase` for more detail.
https://github.com/microsoft/AirSim/blob/7d398add700ce4ba9ebe150dffabd9267f77a0f2/AirLib/src/vehicles/multirotor/api/MultirotorApiBase.cpp#L269-L274
- If there is enough long time where no offboard API calls received, the drone will enter "hover mode" (do not confuse it with the hover API command) for safety, that is, the `goal_mode_` was set to "PositionMode" and the `goal_` was set to the current position. See `OffboardApi::update()` for more detail.
https://github.com/microsoft/AirSim/blob/99208a26e908a156624a8ccfc8aac72029aa75ea/AirLib/include/vehicles/multirotor/firmwares/simple_flight/firmware/OffboardApi.hpp#L43-L59

Now, let me try to explain what happened in your example. 

1. When the first time you set the pose and then hover, everything is OK (although `hover` command did nothing). Suppose the drone was at (x0, y0, z0) at that time. 
2. Then you slept for some time. The drone was entering "hover mode" automatically (in `OffboardApi::update()`) and the `goal_` was set to (x0, y0, z0) and the `goal_mode_` was set to "PositionMode".
3. Then, you set the drone to another position (x1,y1,z1) and then hover, the drone was immediately moved to the point you set, but the `hover` command still did nothing. However, the `goal_` and `goal_mode_` in `OffboardApi` had NEVER been changed! That is why the drone always tries to go back to (x0, y0, z0).
4. The rest is the same as 3.

In short, in your case, the `simSetVehiclePose` and `hover` command NEVER commit any offboard API, the `goal_` was set to the "current position" at the time entering hover mode (that is, when you call `sleep` first time).

To solve this problem, I think we should invoke AT LEAST ONE control command in `moveOnPath` function even the drone was already at the destination (@seheevic also mentioned this problem [here](https://github.com/microsoft/AirSim/issues/1677#issuecomment-553697308)). 

My fixes are committed to this PR, and you can test it with your example. The above reason may be the real cause of the losing-physics problem, so I completely changed my previous submit and rebased all changes. This patch may influence the behavior of `hover`, `moveToZ`, and `moveToPosition` commands.

laurelkeys commented:
Thanks for the detailed explanation @songxiaocheng !

songxiaocheng committed:
Fix losing physics after setting pose

- Set grounded when simSetVehiclePose
- Directly set the vehicle pose, instead of waiting for render.
- Lock for potential conflicts between sim-set-pose and updating physics.

songxiaocheng committed:
Make "moveOnPath" generate at least one drone control command

madratman commented:
/azp run microsoft.AirSim

zimmy87 commented on AirLib/src/vehicles/multirotor/api/MultirotorApiBase.cpp:
```diff
@@ -269,7 +269,7 @@ bool MultirotorApiBase::moveOnPath(const vector<Vector3r>& path, float velocity,
 
     //until we are at the end of the path (last seg is always zero size)
-    while (!waiter.isTimeout() && (next_path_loc.seg_index < path_segs.size()-1 || goal_dist > 0)
+    while (!waiter.isTimeout() && (next_path_loc.seg_index < path_segs.size()-1 || goal_dist >= 0)
```
can you explain the purpose of changing the last conditional to "goal_dist >= 0"?

songxiaocheng replied:
Sorry for my late reply. I made this change in an attempt to resolve the unexpected behavior mentioned above. The detail is in [my comment](https://github.com/microsoft/AirSim/pull/2324#issuecomment-606686250) above. When calling `hover` during the drone is motionless, it invokes `moveOnPath` ultimately, where I think both `next_path_loc.seg_index` and `path_segs.size()-1` are 1. Thus it will never jump in the `while` block with `goal_dist > 0`.

zimmy87 replied:
I tried running @laurelkeys's script with `goal_dist > 0` and `goal_dist >= 0` and I didn't see any change in behavior(the drone still moves toward (0,0,0) after each simSetVehiclePose() call), so I'm not sure this is fully addressing that issue

zimmy87 commented:
Thank you @songxiaocheng for submitting this pull request. I have one open comment I'd like to be addressed before merging.

jonyMarino commented:
Fixes #1246
Fixes #3049

zimmy87 commented:
The bulk of the PR was merged in #3363, with the exception of the one item with an open question
